hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/source/prune/PartitionPruners.java

```diff
@@ -0,0 +1,161 @@
+      this.hivePartition = hivePartition;
+    }
+
+    public Set<String> filter(Collection<String> partitions) {
+      return partitions.stream().filter(this::evaluate).collect(Collectors.toSet());
+    }
+
+    private boolean evaluate(String partition) {
+      String[] partStrArray = FilePathUtils.extractPartitionKeyValues(
+          new org.apache.hadoop.fs.Path(partition),
+          hivePartition,
+          partitionKeys).values().toArray(new String[] {});
+      Map<String, ColumnStats> partStats = new LinkedHashMap<>();
+      for (int idx = 0; idx < partitionKeys.length; idx++) {
+        String partKey = partitionKeys[idx];
+        Object partVal = partKey.equals(defaultParName)
+            ? null : DataTypeUtils.resolvePartition(partStrArray[idx], partitionTypes.get(idx));
+        ColumnStats columnStats = new ColumnStats(partVal, partVal, partVal == null ? 1 : 0);
+        partStats.put(partKey, columnStats);
+      }
+      return partitionEvaluator.stream().allMatch(evaluator -> evaluator.eval(partStats));
+    }
+  }
+
+  /**
+   * Static partition pruner for hoodie table source which partitions list is available in compile phase.
+   * After applied this partition pruner, hoodie source could not read the data from other partitions during runtime.
+   * Note: the data of new partitions created after the job starts would never be read.
+   */
+  public static class StaticPartitionPruner implements PartitionPruner {
+
+    private static final long serialVersionUID = 1L;
+
+    private final Set<String> candidatePartitions;
+
+    private StaticPartitionPruner(Set<String> candidatePartitions) {
+      this.candidatePartitions = candidatePartitions;
+    }
+
+    public Set<String> filter(Collection<String> partitions) {
+      return partitions.stream()
+          .filter(this.candidatePartitions::contains).collect(Collectors.toSet());
+    }
+  }
+
+  /**
+   * Fake partition pruner which would not prune any partitions.
+   */
+  public static class FakePartitionPruner implements PartitionPruner {
+
+    private static final long serialVersionUID = 1L;
+
+    private FakePartitionPruner() {
+
+    }
+
+    @Override
+    public Set<String> filter(Collection<String> partitions) {
+      return new HashSet<>(partitions);
+    }
+  }
+
+  public static PartitionPruner getInstance(
+      List<ExpressionEvaluators.Evaluator> partitionEvaluators,
+      List<String> partitionKeys,
+      List<DataType> partitionTypes,
+      String defaultParName,
+      boolean hivePartition) {
+    if (partitionEvaluators.isEmpty()) {
+      return new FakePartitionPruner();
+    } else {
+      return new DynamicPartitionPruner(partitionEvaluators, partitionKeys, partitionTypes, defaultParName, hivePartition);
+    }
+  }
+
+  public static PartitionPruner getInstance(Collection<String> candidatePartitions) {
+    Set<String> distinctPartitions = new HashSet<>(candidatePartitions);
+    if (distinctPartitions.isEmpty()) {
```

Comment:
In which case the `candidatePartitions` can be empty? Is the `FakePartitionPruner` really needed?